system/security/cryptohelper/pke.cpp

```diff
@@ -0,0 +1,245 @@
+    key.set(EVP_PKEY_new());
+    EVP_PKEY_set1_RSA(key, rsaKey);
+    keyName.set(_keyName);
+}
+
+class CLoadedPublicKeyFromFile : public CLoadedKey
+{
+public:
+    CLoadedPublicKeyFromFile(const char *keyFile, const char *passPhrase)
+    {
+        MemoryBuffer keyMb;
+        loadKeyFileToMem(keyMb, keyFile);
+        loadKeyBio((const char *)keyMb.bytes());
+        RSA *rsaKey = PEM_read_bio_RSA_PUBKEY(keyBio, nullptr, nullptr, (void*)passPhrase);
+        finalize(rsaKey, keyFile);
+    }
+};
+
+class CLoadedPublicKeyFromMemory : public CLoadedKey
+{
+public:
+    CLoadedPublicKeyFromMemory(const char *key, const char *passPhrase)
+    {
+        loadKeyBio(key);
+        RSA *rsaKey = PEM_read_bio_RSA_PUBKEY(keyBio, nullptr, nullptr, (void*)passPhrase);
+        finalize(rsaKey, "<inline>");
+    }
+};
+
+class CLoadedPrivateKeyFromFile : public CLoadedKey
+{
+public:
+    CLoadedPrivateKeyFromFile(const char *keyFile, const char *passPhrase)
+    {
+        MemoryBuffer keyMb;
+        loadKeyFileToMem(keyMb, keyFile);
+        loadKeyBio((const char *)keyMb.bytes());
+        RSA *rsaKey = PEM_read_bio_RSAPrivateKey(keyBio, nullptr, nullptr, (void*)passPhrase);
+        finalize(rsaKey, keyFile);
+    }
+};
+
+class CLoadedPrivateKeyFromMemory : public CLoadedKey
+{
+public:
+    CLoadedPrivateKeyFromMemory(const char *key, const char *passPhrase)
+    {
+        loadKeyBio(key);
+        RSA *rsaKey = PEM_read_bio_RSAPrivateKey(keyBio, nullptr, nullptr, (void*)passPhrase);
+        finalize(rsaKey, "<inline>");
+    }
+};
+
+CLoadedKey *loadPublicKeyFromFile(const char *keyFile, const char *passPhrase)
+{
+    return new CLoadedPublicKeyFromFile(keyFile, passPhrase);
+}
+
+CLoadedKey *loadPublicKeyFromMemory(const char *key, const char *passPhrase)
+{
+    return new CLoadedPublicKeyFromMemory(key, passPhrase);
+}
+
+CLoadedKey *loadPrivateKeyFromFile(const char *keyFile, const char *passPhrase)
+{
+    return new CLoadedPrivateKeyFromFile(keyFile, passPhrase);
+}
+
+CLoadedKey *loadPrivateKeyFromMemory(const char *key, const char *passPhrase)
+{
+    return new CLoadedPrivateKeyFromMemory(key, passPhrase);
+}
+
+// AES
+void aesKeyEncrypt(MemoryBuffer &out, size32_t inSz, const void *inBytes, const char key[aesKeySize], const char iv[aesBlockSize])
+{
+    OwnedCipherCtx ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        throw makeEVPException(0, "Failed EVP_CIPHER_CTX_new");
+
+    /* Initialise the encryption operation. IMPORTANT - ensure you use a key
+     * and IV size appropriate for your cipher
+     * In this example we are using 256 bit AES (i.e. a 256 bit key). The
+     * IV size for *most* modes is the same as the block size. For AES this
+     * is 128 bits
+     * */
+    if (1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, (const unsigned char *)key, (const unsigned char *)iv))
+        throw makeEVPException(0, "Failed EVP_EncryptInit_ex");
```

Comment:
You should log the reason for the failure in all cases that throw
    ERR_error_string(ERR_get_error().....

Reply:
not sure what you mean.. makeEVPException is logging the reason ( ERR_error_string_n(ERR_get_error() ...)